### ISSUE TITLE
display only paragraph text, not headings, in note previews on dashboard and elsewhere

### ISSUE THREAD

**Issue by jywarren** [publiclab/plots2#725]
Hi, this is a `first-timers-only issue`. This means we've worked to make it more legible to folks who either **haven't contributed to our codebase before, or even folks who haven't contributed to open source before**. If you have contributed before, consider leaving this one for someone new, and looking through our general [help wanted](https://github.com/publiclab/plots2/labels/help-wanted) issues. Thanks!

## The Problem

Short, "preview" versions of notes on the site show just the first hundred or so words of a post. But this often shows the first heading, instead of the more useful paragraph text. See for example this page: https://publiclab.org/tag/gsoc

In the line where this shorter version is generated, we should filter out headings before truncating the `node.body` field for display. Currently it looks like this:

```ruby
<p><%= raw strip_tags(sanitize(RDiscount.new(node.body).to_html)).truncate(100) if node.body %></p>
```

And is on this line: https://github.com/publiclab/plots2/blob/master/app/views/notes/_notes.html.erb#L44

## Solution

@david-days has shared a [RegEx "replacement" string](https://en.wikipedia.org/wiki/Regular_expression) that'll get the `body` of a `node` but strip out any Markdown headings (which are written as `## Header name`)

`body.split(/#+.+\n+/,5)[1]` (see [this line](https://github.com/publiclab/plots2/pull/652#discussion-diff-75770666R149) for his usage)

We just need to substitute the above expression for the `node.body` line above, and check that it does in fact remove headings. 

## Steps to Fix

- [ ] claim this issue with a comment here, below
- [ ] set up a repository locally and make sure that all tests pass
- [ ] start a pull request (see contributing to Public Lab software) but mark it as "in progress"
- [ ] fix the issue following the steps above
- [ ] make sure all tests pass with `rake test`
- [ ] alert someone via the developers list (see below) to have your pull request merged. This may take a few extra steps depending on when you do it, but we'll help you out!

Please email the developers list (see https://publiclab.org/developers) if you have questions!

**Comment by nickstaggs** [comment]
I would like to claim this, is there a timeline you guys are looking for? I could definitely complete it in the next week but I can expedite the process and finish by the end of the weekend if need be.

**Comment by jywarren** [comment]
Hi, @nickstaggs - great to hear it -- no, no urgency here. We're glad to have your contribution and please reach out if you have any trouble. Thanks!

**Comment by david-days** [comment]
@nickstaggs - If you have any issues with the RegEx, let me know.  I'm not an expert on regex, but I've been working with it long enough to figure out how to (crudely) get what I need.

If you come up with any improvements or see a better way to do this, I'm all ears--having a good preview of content from the database could be used in lots of different places.

**Comment by nickstaggs** [comment]
Anytime @jywarren - could you assign the issue to me. @david-days - thanks a lot I'll let you know if I run into any issues or if I think of anyways to optimize the RegEX. I'm going to dive in and give you guys a status report at the end of the weekend.

**Comment by jywarren** [comment]
sure -- doing that now. 

You could also -- just as a possibility -- make a "preview" method in the `/app/models/drupal_node.rb` file which makes the truncated version available to everyone:

```ruby
def body_preview # or maybe "body_noheaders"?
  self.body.split(/#+.+\n+/,5)[1]
end
```

or get fancy and have a length parameter:

```ruby
def body_preview(length = 100) # 100 default
  self.body.split(/#+.+\n+/,5)[1].truncate(length)
end
```

Just some ideas, hope they're helpful!

**Comment by jywarren** [comment]
Ah, i think to assign to you, you have to be a collaborator, so i sent you an invite. Not totally necessary but happy to do it if you click "accept" on the invite. Thanks!

**Comment by nickstaggs** [comment]
Hey guys, hope you had good weekend. A few things:
1. Could I please get the permissions to fetch from upstream to keep my clone up to date?
2. I was wondering what the output to rake test:all is supposed to look like, I got this

237 tests, 554 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
0% passed

or something very similar for all the sections, there were no failures or errors but I am a little worried about the 0% passed.

3. I think I will have to make a preview method because the body.split(/#+.+\n+/,5)[1] will return nil if the body does not start with a header.

I will get the pull request started tomorrow or Tuesday and I will keep you guys informed on how the preview method is going.

**Comment by jywarren** [comment]
Hi, Nick - I don't think you need special permission to fetch from upstream
master -- just `git pull https://github.com/publiclab/plots2.git master` --

as to the tests, you'll need phantomjs to run the final (javascript) tests,
as described here:
https://github.com/publiclab/plots2/#phantomjs-for-javascript-tests

However, as your code doesn't touch JavaScript, I wouldn't worry about
running those. Your Rails tests work fine, and the JS tests will be run
automatically on our test server when you submit your PR, so we'll have a
check then.

If you make a model method, and you're feeling ambitious, you could add a
unit test to this file (
https://github.com/publiclab/plots2/blob/master/test/unit/drupal_node_revision_test.rb)
which scans the output of the method you write for `<h1>` or `<h2>` header
tags. I'm also happy to provide some guidance on unit testing. This could
let you test cases like not having a header at all.

Thanks!

**Comment by jywarren** [comment]
Yes, this was solved in #788! Thanks, closing now, sorry!

**Comment by ebarry** [comment]
Hi @nickstaggs! Thank you again for your valuable contribution to Public Lab. I would like to invite you to join us in a video conference to say a few words about what you contributed and how your experience was. The event is https://publiclab.org/openhour, and it’s coming right up on November 7th, at 1pm EDT which is UTC/GMT -4. 

Please feel free to respond here, or else email liz@publiclab.org, stevie@publiclab.org, jeff@publiclab.org.
